nixos/modules/services/networking/dhcpcd.nix

```diff
@@ -77,20 +77,21 @@ let
       ${cfg.extraConfig}
     '';
 
-  exitHook = pkgs.writeText "dhcpcd.exit-hook"
-    ''
+  exitHook = pkgs.writeText "dhcpcd.exit-hook" ''
+    ${optionalString (config.services.ntp.enable || config.services.openntpd.enable || config.services.chrony.enable) ''
```

Comment:
You check these now in multiple places and they have to be kept in sync. This is usually error-prone in the long run.

I would rather have the hook generated unconditionally and this `if` statement executed unconditionally. In the worst case, the `if` statement will be empty, but that's pretty harmless. But it cuts down on possibilities that the whole construct could be wrong in some configuration.

So what I would suggest is:

- drop the `mkIf` for  `environment.etc."dhcpcd.exit-hook"`
- drop the `optionalString` that wraps everything here
- keep the `optionalString config.services.*.enable` to avoid generating pointless warnings in the log.

Reply:
In this case, by default, the following script will be generated  with an empty action.

Reply:
Tried to simplify the script.

Reply:
Updated.

Reply:
> In this case, by default, the following script will be generated with an empty action.

If you look at /etc you'll see this is very common: there's no effort to generate good-looking files in NixOS. It would be good for manual inspection of a generation, but in general it's too much work for too little value.

Reply:
> So what I would suggest is:
> 
> * drop the `mkIf` for  `environment.etc."dhcpcd.exit-hook"`
> * drop the `optionalString` that wraps everything here
> * keep the `optionalString config.services.*.enable` to avoid generating pointless warnings in the log.

Updated.

Reply:
The last changes bring back the redundant checks. What did I miss?